qa/tasks/util/chacra.py

```diff
@@ -0,0 +1,186 @@
+
+    if len(result) == 0:
+        raise RuntimeError(f'no results found at {resp.url}')
+
+    # TODO: filter the result down to the correct arch etc.?
+    result = result[0]
+
+    status = result['status']
+    if status != 'ready':
+        raise RuntimeError(f'cannot pull file with status: {status}')
+
+    # build the chacra url
+    chacra_host = urlparse(result['url']).netloc
+    chacra_ref = result['ref']
+    chacra_sha1 = result['sha1']
+    log.info(f'got chacra host {chacra_host}, ref {chacra_ref}, sha1 {chacra_sha1} from {resp.url}')
+
+    # prefer codename if a release is not specified
+    if result.get('distro_codename'):
+        release = result.get('distro_codename')
+    elif result.get('distro_version'):
+        release = result.get('distro_version')
+    elif not release:
+        raise RuntimeError('cannot determine distro release!')
+
+    if not arch:
+        if ARCH in result['archs']:
+            arch = ARCH
+        elif len(result['archs']) > 0:
+            arch = result['archs'][0]
+        else:
+            raise RuntimeError('cannot determine the arch type!')
+
+    # build the url to the binary
+    url = _get_binary_url(
+        chacra_host,
+        result['project'],
+        chacra_ref,
+        chacra_sha1,
+        result['distro'],
+        release,
+        arch,
+        result['flavor'],
+        filename,
+    )
+
+    return url
+
+def pull(
+    filename,
+    project=None,
+    distro=None,
+    release=None,
+    arch=None,
+    flavor=None,
+    branch=None,
+    sha1=None
+):
+    '''
+    Pull a build result from chacra
+    '''
+    url = get_binary_url(
+            filename,
+            project=project,
+            distro=distro,
+            release=release,
+            arch=arch,
+            flavor=flavor,
+            branch=branch,
+            sha1=sha1
+    )
+    resp = requests.get(url, stream=True)
+    resp.raise_for_status()
+    log.info(f'got file from {resp.url}')
+
+    return resp
+
+def main():
+    handler = logging.StreamHandler(sys.stdout)
+    log.addHandler(handler)
+    log.setLevel(logging.INFO)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--project', default=PROJECT)
+    parser.add_argument('--distro', default=DISTRO)
+    parser.add_argument('--release', default=RELEASE)
+    parser.add_argument('--arch', default=ARCH)
+    parser.add_argument('--branch', default=BRANCH)
+    parser.add_argument('--sha1', default=SHA1)
+    parser.add_argument('--flavor', default=FLAVOR)
+    parser.add_argument('--src', default=FILENAME)
+    parser.add_argument('--dest', default=FILENAME)
+    args = parser.parse_args()
+
+    resp = pull(
+        args.src,
+        project=args.project,
+        distro=args.distro,
+        release=args.release,
+        arch=args.arch,
+        flavor=args.flavor,
+        branch=args.branch,
+        sha1=args.sha1
+    )
+
+    dest = Path(args.dest).absolute()
+    with open(dest, 'wb') as f:
```

Comment:
you may want to set correct permissions when writing a binary file.

Reply:
Thanks for pointing that out, however I think the permissions are being adjusted later in the code. It could make a good future enhancement. However, since this PR is long and complex I don't want to make that change now since the tests *are* passing.